File: apps/web/src/api/schemas.ts
import { z } from 'zod';

export const deviceSchema = z.object({
  id: z.string(),
  name: z.string(),
  serialNumber: z.string(),
  firmwareVersion: z.string(),
  hardwareType: z.string(),
  connectionType: z.enum(['http', 'mqtt', 'gateway']),
  status: z.enum(['online', 'offline', 'warning']),
  location: z.string(),
  lastSeenAt: z.string()
});

export const telemetryPointSchema = z.object({
  deviceId: z.string(),
  metric: z.string(),
  value: z.number(),
  unit: z.string(),
  recordedAt: z.string()
});

export const eventSchema = z.object({
  id: z.string(),
  deviceId: z.string(),
  type: z.string(),
  message: z.string(),
  severity: z.enum(['info', 'warning', 'critical']),
  createdAt: z.string()
});

export const ruleSchema = z.object({
  id: z.string(),
  deviceId: z.string(),
  metric: z.string(),
  operator: z.enum(['>', '>=', '<', '<=', '=']),
  threshold: z.number(),
  enabled: z.boolean()
});

export const devicesSchema = z.array(deviceSchema);
export const telemetrySchema = z.array(telemetryPointSchema);
export const eventsSchema = z.array(eventSchema);
export const rulesSchema = z.array(ruleSchema);

export type Device = z.infer<typeof deviceSchema>;
export type TelemetryPoint = z.infer<typeof telemetryPointSchema>;
export type DeviceEvent = z.infer<typeof eventSchema>;
export type Rule = z.infer<typeof ruleSchema>;

File: apps/api/app/schemas.py
from datetime import datetime
from enum import StrEnum
from typing import Annotated

from pydantic import BaseModel, Field, field_validator


NonEmptyString = Annotated[str, Field(min_length=1)]


class ConnectionType(StrEnum):
    HTTP = "http"
    MQTT = "mqtt"
    GATEWAY = "gateway"


class DeviceStatus(StrEnum):
    ONLINE = "online"
    OFFLINE = "offline"
    WARNING = "warning"


class EventSeverity(StrEnum):
    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


class RuleOperator(StrEnum):
    GT = ">"
    GTE = ">="
    LT = "<"
    LTE = "<="
    EQ = "="


class DeviceCreate(BaseModel):
    name: NonEmptyString
    serialNumber: NonEmptyString
    firmwareVersion: NonEmptyString = "0.1.0"
    hardwareType: NonEmptyString
    connectionType: ConnectionType = ConnectionType.HTTP
    location: NonEmptyString


class Device(DeviceCreate):
    id: NonEmptyString
    status: DeviceStatus
    lastSeenAt: datetime | None = None
    createdAt: datetime
    updatedAt: datetime


class DeviceCreateResponse(Device):
    apiKey: NonEmptyString


class TelemetryReadingIn(BaseModel):
    metric: NonEmptyString
    value: float
    unit: NonEmptyString

    @field_validator("metric")
    @classmethod
    def metric_must_be_machine_safe(cls, value: str) -> str:
        if not value.replace("_", "").replace("-", "").isalnum():
            raise ValueError("metric must use letters, numbers, dash, or underscore")
        return value


class TelemetryIngestRequest(BaseModel):
    deviceId: NonEmptyString
    firmwareVersion: NonEmptyString
    timestamp: datetime
    readings: list[TelemetryReadingIn] = Field(min_length=1, max_length=64)


class TelemetryPoint(BaseModel):
    id: NonEmptyString
    deviceId: NonEmptyString
    metric: NonEmptyString
    value: float
    unit: NonEmptyString
    recordedAt: datetime
    receivedAt: datetime


class TelemetryIngestResponse(BaseModel):
    deviceId: NonEmptyString
    accepted: int
    receivedAt: datetime


class DeviceEvent(BaseModel):
    id: NonEmptyString
    deviceId: NonEmptyString
    type: NonEmptyString
    message: NonEmptyString
    severity: EventSeverity
    createdAt: datetime


class Rule(BaseModel):
    id: NonEmptyString
    deviceId: NonEmptyString
    metric: NonEmptyString
    operator: RuleOperator
    threshold: float
    enabled: bool
